lib/trino-collect/src/main/java/io/trino/collect/cache/EvictableCache.java

```diff
@@ -105,7 +105,10 @@ public V get(K key, Callable<? extends V> valueLoader)
         Token<K> newToken = new Token<>(key);
         Token<K> token = tokens.computeIfAbsent(key, ignored -> newToken);
         try {
-            return dataCache.get(token, valueLoader);
+            V value = dataCache.get(token, valueLoader);
+            // revive token if it got expired before being loaded again
+            tokens.computeIfAbsent(key, ignored -> token);
```

Comment:
At this point, it's not guaranteed that `dataCache` still has the `(token, value)` entry.
Of course, during `V value = dataCache.get(token, valueLoader);` it existed, but between these two lines, it could got expired/invalidated/evicted due to size limits, etc.
in such case, our token won't be removed ever again, and we would have a memory leak

Reply:
From what I understand, eviction due to any of those rules happens before loading new values, and if `get()` returned a value that value should be considered fresh and not subject to removal.

Reply:
Hmm, yea you're right. Unfortunately it doesn't look like the cache allows you to know for sure whether an expired item is replaced within the same call.